nx/guides/modoc_agg_nx_01.livemd

```diff
@@ -0,0 +1,826 @@
+# Livebook companion for Nx
+
+```elixir
+Mix.install([
+  {:nx, "~> 0.4.2"},
+  {:kino, "~> 0.8"}
+])
+
+# import IEx.Helpers
+```
+
+## Aggregations operations on tensors
+
+
+
+```elixir
+import Nx, only: :sigils
+```
+
+A tensor _in computer science_ can be viewed as a list of multi-indexed numbers. You simply wrap your list with `Nx.tensor`. Equivalently, you can use the compact `~V` sigil:
```

Comment:
I at last took time to read, sorry for the delay. 

Just a remark. The word tensor is ambiguous. It is related to the math we learned at school by the fact that it is multi-indexed (I believe "multi" is important), but that's all.
Ex. No intention of being pedantic at all, but the rank in computer science is just the number of dimensions which of course is not the case in math. Also, you agree that the (math) transpose of a covector (linear form, covariant) is a vector (contravariant), and vice-versa. Here, no, as per the definition of a vector. so I was confused because a covector has shape {n,1}, thus is 2D in computer science whilst not in math.  But again, it's just a matter of definition related to the domain and not important of course.

Maybe some examples to introduce how to build tensors are welcome? Such as:

 <==
We use the function `Nx` to construct tensors. A few examples:
 1. You can pass a number to a tensor and you get a 0D tensor. For example, `Nx(1)` is a 0D tensor.
2. you can build a 1D tensor or "vector" by passing a list of numbers. For example, `Nx([1,2,3])`.  Equivalently, you can use the sigil `~V` as `~V(1 2 3)`. 
3. you can build a matrix or 2D tensor with the sigil `~M`. 
For example a (2,3) matrix
```elixir
~M( 1 2 3
4 5 6)
```
Equivalently, we can use nested lists: `Nx.tensor([[1, 2,3],[4,5,6]])`.
Equivalently, we can reshape a vector: `~V(1 2 3 4 5 6) |>  Nx.reshape({2,3})`.

4. You can pass an arbitrarily nested list of numbers. For example, `Nx.tensor([[[1,2],[3,4]],[[5,6],[7,8]]]))`is a 3D tensor of shape `{2,2,3}`.  Equivalently, you can reshape a vector: `~V(1 2 3 4 5 6 7 8) |> Nx.Reshape({2,2,2})`.
 ==>

Reply:
I think this whole "what is and how to create a tensor" could be its own guide instead of staying in this one.

> Just a remark. The word tensor is ambiguous. 
> It is related to the math we learned at school by the fact that it is multi-indexed (I believe "multi" is important), but that's all.

Yes, but here we are _defining_ it. Therefore it's not ambiguous anymore in our domain. We can make it more clear by prefacing with "Tensors have many different definitions. In Nx, a tensor is..." or something like that.
I don't believe multi is important, but it's not a hill I'm willing to die on :)

> Ex. No intention of being pedantic at all, but the rank in computer science is just the number of dimensions which of course is not the case in math. 
Yes, but we have to again remember that we're speaking in the confines of the Nx domain language. No need to confuse people by introducing math-speak. This ties in to the next point:

> Also, you agree that the (math) transpose of a covector (linear form, covariant) is a vector (contravariant), and vice-versa. Here, no, as per the definition of a vector. so I was confused because a covector has shape {n,1}, thus is 2D in computer science whilst not in math. But again, it's just a matter of definition related to the domain and not important of course.

None of these definitions really matter to Nx. We can define that a vector is a 1D tensor just for the sake of speaking more easily about things, like we did with matrices being 2D tensors. In general, I think you're coming into this with a "math" mindset. This isn't bad in itself, but you have to remember that the ubiquitous language changes with each domain, and in Nx-land, we don't have the same definitions that you might be used to.

All in all, I think all of this might have a place still by discussing differences and similarities in a more in-depth _separate_ guide. What do you think?

Reply:
In computer science, I have only used numpy or sagemth, so code running on CPU, never tensors on GPU! But I agree, it's just a matter of definition.